scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/PostgresSpecificationRepository.java

```diff
@@ -89,21 +91,23 @@ private static GoalCompilationResult compileGoalDefinition(
       final PostgresGoalRecord pgGoal,
       final SchedulingDSLCompilationService schedulingDSLCompilationService)
   {
-    final SchedulingDSL.GoalSpecifier goalSpecifier;
-    try {
-      goalSpecifier = schedulingDSLCompilationService.compileSchedulingGoalDSL(
-          planId,
-          pgGoal.definition(),
-          pgGoal.name());
-    } catch (SchedulingDSLCompilationService.SchedulingDSLCompilationException e) {
-      return new GoalCompilationResult.Failure(e.getMessage());
+    final var goalCompilationResult = schedulingDSLCompilationService.compileSchedulingGoalDSL(
+        planId,
+        pgGoal.definition(),
+        pgGoal.name());
+
+    if (goalCompilationResult instanceof SchedulingDSLCompilationService.SchedulingDSLCompilationResult.Error g) {
+      return new GoalCompilationResult.Failure(g.errors());
+    } else if (goalCompilationResult instanceof SchedulingDSLCompilationService.SchedulingDSLCompilationResult.Success g) {
```

Comment:
I was thinking the below is really wordy. Thoughts?
 `SchedulingDSLCompilationService.SchedulingDSLCompilationResult.Success` 

to

`SchedulingDSLCompilationService.results.Success`

Reply:
I'm hesitant about having it as just `Result` because I think that has some very specific connotations w.r.t. shape and behavior. IE a standard `Result` type is more of the style `Result<SuccessType, ErrorType>` with associated methods to act on it. I'm entirely open to changing it though if there is consensus otherwise (I might be polluted by my affection for the Rust Result type here)

Reply:
That makes sense that others might see the `result` type being `Result<SuccessType, ErrorType>`

Maybe `SchedulingDSLCompilationService.CompilationResult.Success` but this is in the weeds and was just an idea I had so if no one and any strong feelings you can ignore this. :)

Reply:
You're making me feel that we need a Result utility type for Java. But I know that java is very throw - encouraging, so don't feel I have sufficient background to make that suggestion. 

Personally, I've very throw-averse (probably from too much time with Rust)

Reply:
> You're making me feel that we need a Result utility type for Java.

We kinda sorta have one in `gov.nasa.jpl.aerie.merlin.framework.Result`, but it was developed to meet model use-cases; I'd argue against just having other components import the modeling `framework`. Anyway, it could use some touching-up since it was created before we had `sealed` interface -- but we don't yet have pattern-matching in `switch` to seal the deal.

Reply:
> Personally, I've very throw-averse (probably from too much time with Rust)

Personally, I agree with you, but Java makes it pretty painful to avoid exceptions. I tend to categorize exceptions as one of three kinds: caller-callee contract exceptions, component integration exceptions, and we-didn't-code-for-this-scenario exceptions.

Contract exceptions occur when the caller didn't respect some requirements that the callee needs for correct operation, e.g. indices out of bounds and whatnot. These ought to be checked exceptions, which as far as the calling convention is concerned makes them like additional variants of a returned sum type (that you happen to match on using `try-catch`). Sometimes it can make sense to prevent this kind of exception from capturing call stack information (since that's somewhat performance-sensitive); you can do this by passing the appropriate `bool` flags to the exception's superclass in the constructor.

Integration exceptions occur when code can't abide by its contract due to reasons outside its control, i.e. because some other component it has been integrated with (such as via dependency injection) has failed. These especially happen when the callee's detailed contract cannot be communicated to the caller, e.g. because it's hidden behind an abstract interface that suppresses some of those details. Since the caller can't be expected to abide by requirements it can't be aware of, responsibility falls to the agent that integrated these components. (This is a very Erlang supervision-like perspective on failure.)

"We-didn't-code-for-this-scenario exceptions" are as they say: some design assumptions have been violated, so the program simply can't have a strategy for handling this in the first place. These are like top-level integration exceptions in that the developer is at fault for wiring together these components, but it's a little bit broader of a category. It's most like Rust's `panic`.

Reply:
This is exactly the write up that I went hunting for because I thought @Twisol had written it before. Copied and saved this now for prosperity. :)

Reply:
I did write it somewhere, a long time ago in another PR, but good luck finding it ^_^;